dev_setup.sh

```diff
@@ -0,0 +1,210 @@
+# tested.
+#
+
+set -e
+
+HERE=$(cd `dirname $0`; pwd)
+LOG=$HERE/dev_setup.log
+ROOT=$HERE
+
+# Create a new, empty file for environment variables
+VAR_FILE=$ROOT/env.sh
+truncate -s 0 $VAR_FILE
+echo "# Generated environment variables. Manual changes will get clobbered by dev_setup.sh" >> $VAR_FILE
+
+WHAT4_SOLVERS_SNAPSHOT="snapshot-20240212"
+WHAT4_SOLVERS_URL="https://github.com/GaloisInc/what4-solvers/releases/download/$WHAT4_SOLVERS_SNAPSHOT/"
+WHAT4_SOLVERS_MACOS_12="macos-12-X64-bin.zip"
+WHAT4_SOLVERS_MACOS_14="macos-14-ARM64-bin.zip"
+WHAT4_CVC4_VERSION="version 1.8"
+WHAT4_CVC5_VERSION="version 1.1.1"
+
+# Set of supported platforms:
+MACOS14="macos14"
+MACOS12="macos12" # actually, this isn't supported yet
+
+USED_BREW=false
+
+# Returns a string indicating the platform (from the set above), or empty if
+# a supported platform is not detected.
+function supported_platform {
+    # Make sure we're running on a supported platform
+    case $(uname -s) in
+        Darwin)
+            if [ $(uname -m) = "arm64" ]; then
+                echo $MACOS14
+            # This is how we'd support macOS 12. Since this hasn't been tested yet,
+            # we withhold official support.
+            # This might bork on something running macOS <12, since we're basing
+            # the it on the hardware, not the specific version.
+            elif [ $(uname -m) = "x86_64" ]; then
+                echo ""
+            fi;;
+        *) echo ""
+    esac
+}
+
+RED="\033[0;31m"
+function notice {
+    echo "[NOTICE] $*"
+}
+
+function is_installed {
+    command -v "$@" >/dev/null 2>&1
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        # This may or may not be the same directory that this script lives in.
+        mkdir -p `dirname $LOG`
+        echo "$@" >>$LOG
+        if ! "$@" >>$LOG 2>&1
+        then
+            echo
+            tail -n 50 $LOG
+            echo
+            echo "[ERROR] An error occurred; please see $LOG"
+            echo "[ERROR] The last 50 lines are printed above for convenience"
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $ROOT
+    notice "Updating submodules"
+    logged git submodule update --init
+}
+
+function install_ghcup {
+    if ! is_installed ghcup; then
+        notice "Installing GHC toolchain via GHCup"
+        # Manually install Haskel toolchain. If this doesn't work, try
+        # using the install script at https://www.haskell.org/ghcup/
+        if [ $CRYPTOL_PLATFORM = $MACOS12 ] || [ $CRYPTOL_PLATFORM = $MACOS14 ]; then
+            logged brew install ghcup
+            USED_BREW=true
+        else
+            notice "Did not install GHCup. This script only supports macOS 14"
+        fi
+
+        logged ghcup install ghc --set recommended
+        logged ghcup install cabal --set recommended
+        logged ghcup install stack --set recommended
+        logged ghcup install hls --set recommended
```

Comment:
I think we should avoid referencing ghcup's *recommended* versions here. The recommended version will change over time and won't necessarily be in sync with the version we intended to target. We could list at the top of this script the versions of tools we require and then install exactly those versions with these commands. When it comes up invoking the versions of things we want it can be done explicitly by version.

For example when building with cabal we can do: `cabal configure -w ghc-9.6.5` to explicitly set the build to use the expected version of GHC. If we want to lock down the version of cabal we use we could also go further: `cabal-3.10.3.0 configure -w ghc-9.6.5`

Reply:
As far as I know we don't use stack in our builds. We probably don't need to install it as part of this development environment.

Reply:
+1 on locking down GHC and related tools here.

Reply:
There are cabal configs for three GHC versions, though; which one? Or should we install all three by default?

Reply:
Yeah... good question. ISTM having this script use those cabal configs (or whatever canonically defines which GHCs we need to support) to find out which GHC versions to install might be nice, in the spirit of having one source of truth for that.